src/main/java/org/truffleruby/core/module/ModuleFields.java

```diff
@@ -323,24 +324,27 @@ public void setAutoloadConstant(RubyContext context, Node currentNode, String na
         setConstantInternal(context, currentNode, name, filename, true);
     }
 
-    public void setConstantInternal(RubyContext context, Node currentNode, String name, Object value, boolean autoload) {
+    public boolean setConstantInternal(RubyContext context, Node currentNode, String name, Object value, boolean autoload) {
         checkFrozen(context, currentNode);
-        // TODO(CS): warn when redefining a constant
-        // TODO (nirvdrum 18-Feb-15): But don't warn when redefining an autoloaded constant.
 
         SharedObjects.propagate(context, rubyModuleObject, value);
 
+        boolean redefinedNotAutoload = false;
         while (true) {
             final RubyConstant previous = constants.get(name);
             final boolean isPrivate = previous != null && previous.isPrivate();
             final boolean isDeprecated = previous != null && previous.isDeprecated();
             final RubyConstant newValue = new RubyConstant(rubyModuleObject, value, isPrivate, autoload, isDeprecated);
 
             if ((previous == null) ? (constants.putIfAbsent(name, newValue) == null) : constants.replace(name, previous, newValue)) {
+                if (!autoload && previous != null) {
+                    redefinedNotAutoload = true;
+                }
                 newConstantsVersion();
                 break;
```

Comment:
You could just `return !autoload && previous != null` here.

Reply:
Fixed.